src/main/java/com/telefonica/iot/cygnus/backends/mongo/MongoBackend.java

```diff
@@ -466,4 +473,91 @@ private Date getOrigin(GregorianCalendar calendar, Resolution resolution) {
         return new Date(gc.getTimeInMillis());
     } // getOrigin
     
+    /**
+     * Stores in per-service/database "collection_names" collection the matching between a hash and the fields used to
+     * build it.
+     * FIXME: destination is under study
+     * @param dbName
+     * @param hash
+     * @param isAggregated
+     * @param fiwareService
+     * @param fiwareServicePath
+     * @param entityId
+     * @param entityType
+     * @param attrName
+     * @param destination
+     * @throws java.lang.Exception
+     */
+    public void storeCollectionHash(String dbName, String hash, boolean isAggregated, String fiwareService,
+            String fiwareServicePath, String entityId, String entityType, String attrName, String destination)
+        throws Exception {
+        // get the database and the collection; the collection is created if not existing
+        MongoDatabase db = getDatabase(dbName);
+        MongoCollection collection;
+
+        try {
+            LOGGER.debug("Creating Mongo collection=collection_names at database=" + dbName);
+            db.createCollection("collection_names");
+        } catch (Exception e) {
+            if (e.getMessage().contains("collection already exists")) {
+                LOGGER.debug("Collection already exists, nothing to create");
+            } else {
+                throw e;
+            } // if else
+        } finally {
+            collection = db.getCollection("collection_names");
+        } // try catch finally
+
+        // Two updates operations are needed since MongoDB currently does not support the possibility to address the
+        // same field in a $set operation as a $setOnInsert operation. More details:
+        // http://stackoverflow.com/questions/23992723/ \
+        //     findandmodify-fails-with-error-cannot-update-field1-and-field1-at-the-same
+        
+        // build the query
+        BasicDBObject query = new BasicDBObject().append("_id", hash + (isAggregated ? ".aggr" : ""));
+        
+        // do the first update
+        BasicDBObject update = buildUpdateForCollectionHash("$setOnInsert", isAggregated, fiwareService,
+                fiwareServicePath, entityId, entityType, attrName, destination);
+        LOGGER.debug("Updating data, database=" + dbName + ", collection=collection_names, query="
+                + query.toString() + ", update=" + update.toString());
+        UpdateResult res = collection.updateOne(query, update, new UpdateOptions().upsert(true));
+/*
+        Theoretically, this part is needed, but it works with a single unpdate operation!
```

Comment:
That's true, in fact the hash is only stored when the collection storing the data is created and that is done hopefully once ;)

Related to this, I highly recommend including a tech-debt issue regarding the fact that we should be controlling hash collisions and acting upon it. After discussing it offline, we decide to log the hash collision as a warning and do not store raw or aggregated data since that may generate wrong results for the aggregated data (in case the attribute name of the entities causing the collision also had the same name) ;)